plugin-server/src/worker/ingestion/event-pipeline/embedErrorEventStep.ts

```diff
@@ -0,0 +1,82 @@
+import { FeatureExtractionPipeline, FeatureExtractionPipelineOptions } from '@xenova/transformers'
+
+import { defaultConfig } from '../../../config/config'
+import { runInstrumentedFunction } from '../../../main/utils'
+import { PreIngestionEvent } from '../../../types'
+import { status } from '../../../utils/status'
+import { EventPipelineRunner } from './runner'
+
+const options: FeatureExtractionPipelineOptions = { pooling: 'mean', normalize: false }
+
+const errorEmbeddingModel = defaultConfig.ERROR_EVENT_EMBEDDING_MODEL
+
+let featureExtractionPipeline: FeatureExtractionPipeline | null
+
+// this downloads the model the first time it is run, so we do this on server start-up
+// we only want to spend the cost of loading the model if we are actually using it
+export async function initEmbeddingModel(
+    anyTeamsAreEmbeddingEvents: boolean
+): Promise<FeatureExtractionPipeline | null> {
+    try {
+        if (!featureExtractionPipeline && anyTeamsAreEmbeddingEvents) {
+            // a little magic here to both delay the slow import until the first time it is needed
+            // and to make it work due to some ESM/commonjs faff
+            status.info('🤖', 'initialising the error embedding model')
+            await runInstrumentedFunction({
+                func: async () => {
+                    const TransformersApi = Function('return import("@xenova/transformers")')()
+                    const { pipeline } = await TransformersApi
+                    featureExtractionPipeline = await pipeline('feature-extraction', errorEmbeddingModel)
```

Comment:
This will download [several hundreds of megabytes](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1/tree/main/onnx) everytime, and we cannot have ingestion going down when the download server has an incident.

Definitely needs to be sourced from S3 or other internal storage, or baked in the docker image: @frankh what would be your recommendation here? Ingress is technically free, but we'll put strain on the NAT gateways.

Reply:
I can put this in our S3 by hand for now and source from there somehow

I was trying to avoid loading on startup so I can get to a place I can test the embeddings away from my local machine. I don't want to invest too much here since we might not end up using it.

Reply:
ok, so the library will load from local file or from a configured remote host (instead of hugging face).

i guess i can mirror the structure per model in S3 with cloud front in front of it 

so something like this in the code so we only support this when the remote host is configured

```
const errorEmbeddingModel = defaultConfig.ERROR_EVENT_EMBEDDING_MODEL
if (defaultConfig.ERROR_EMBEDDING_REMOTE_HOST === null) {
    // this effectively disables this feature
    env.allowRemoteModels = false
    env.allowLocalModels = false
} else {
    env.remoteHost = defaultConfig.ERROR_EMBEDDING_REMOTE_HOST
}
```

@frankh I'm assuming S3 with cloudfront is the easiest way? I can do terrible terraform to upset you in the morning 🙈

Reply:
No need for cloudfront caching for internal uses, I'd recommend terraforming a S3 bucket we can write to from pods, and writing a simple [customCache](https://github.com/xenova/transformers.js/blob/da2688626d7812ad1ea47fd304c2072cc685051b/src/env.js#L116C5-L116C16) implem that uses S3 (there might be one on npm already)

The first pods to come in will seed the cache for the other pods to reuse. When changing the model, we can change it on the overflow deploy first, to avoid 100 pods thundering to update the cache

Reply:
If we go this route, let's keep the bucket name generic about all ingestion concerns, because we'll most probably use it to store the mmdb database, instead of pulling it from PG on every pod.

Reply:
some discussion here https://posthog.slack.com/archives/C045QJCUGQ4/p1718188225696009

i'll switch to doing more local testing (although I'm out today and tomorrow) while that progresses